pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -96,192 +158,374 @@ func shouldRestartEncryptionKeyRotation(cp *rkev1.RKEControlPlane) bool {
+		if err != nil {
+			return err
+		}
+	}
+	var (
+		leaderStage string
+		err         error
+	)
+	switch rke2.GetRuntime(cp.Spec.KubernetesVersion) {
+	case rke2.RuntimeRKE2:
+		leaderStage, err = p.encryptionKeyRotationRestartRKE2(cp, clusterPlan, leader)
+	case rke2.RuntimeK3S:
+		leaderStage, err = p.encryptionKeyRotationRestartK3S(cp, clusterPlan, leader)
 	}
-	err = assignAndCheckPlan(p.store, fmt.Sprintf("encryption key rotation: [%s]", cp.Status.RotateEncryptionKeysPhase), leader, nodePlan, 0, 0)
 	if err != nil {
-		return p.enqueueIfErrWaiting(cp, err)
+		return err
 	}
 
-	err = p.encryptionKeyRotationUpdateControlPlanePhase(cp, rkev1.RotateEncryptionKeysPhaseVerifyLeaderStatus)
+	// a phase has been completed at least once for this generation
+	if leaderStage == encryptionKeyRotationStageReencryptFinished &&
+		cp.Status.RotateEncryptionKeys.Stage != "" {
+		return p.encryptionKeyRotationFinished(cp)
+	}
+
+	cp.Status.RotateEncryptionKeys.Stage = leaderStage
+	err = p.encryptionKeyRotationUpdateControlPlanePhase(cp, rkev1.RotateEncryptionKeysPhaseApplyLeader)
 	if err != nil {
 		return err
 	}
 	return generic.ErrSkip
 }
 
-func (p *Planner) encryptionKeyRotationVerifyLeaderStatus(cp *rkev1.RKEControlPlane, leader *planEntry) error {
-	return p.encryptionKeyRotationGetLeaderStatus(cp, leader, rkev1.RotateEncryptionKeysPhaseRestartFollowers)
+// encryptionKeyRotationIsControlPlaneAndNotLeader allows us to filter cluster plans to restart healthy follower nodes.
+func encryptionKeyRotationIsControlPlaneAndNotLeader(cp *rkev1.RKEControlPlane) roleFilter {
+	return func(entry *planEntry) bool {
+		return encryptionKeyRotationIsSuitableControlPlane(entry) &&
+			cp.Annotations[encryptionKeyRotationLeaderAnnotation] != entry.Machine.Name
+	}
 }
 
-// encryptionKeyRotationRestartFollowers restarts each follower, and ensures that the status matches the leader node.
-// If this is the first time this function has been run this generation, the generation status will be updated, however
-// the phase will not be set to done unless the controlplane secrets-encrypt status is uniformly reencrypt_finished, as
-// subsequent runs will begin in the reencrypt_finished phase, and require the applyLeader phase to be reached at least
-// once before completion.
-func (p *Planner) encryptionKeyRotationRestartFollowers(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan, leader *planEntry) error {
-	leaderPhase, err := encryptionKeyRotationPhaseFromPeriodic(leader)
-	if err != nil {
-		return p.restartEncryptionKeyRotation(cp)
+// encryptionKeyRotationRestartRKE2 restarts the nodes using encryptionKeyRotationRestartServices for an HA cluster, and
+// encryptionKeyRotationIsSuitableControlPlane if there is only one controlplane node.
+func (p *Planner) encryptionKeyRotationRestartRKE2(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan, leader *planEntry) (string, error) {
+	if len(collect(clusterPlan, encryptionKeyRotationIsSuitableControlPlane)) == 1 {
```

Comment:
nit, and you can get rid of the switch statement above:
```suggestion
	if rke2.GetRuntime(cp) == rke2.RuntimeRKE2 && len(collect(clusterPlan, encryptionKeyRotationIsSuitableControlPlane)) == 1 {
```